### PR TITLE
feat(builder): add new webpack progress bar

### PR DESCRIPTION
**Pull request by zhusjfaker**
# PR Details

## Description

* create a new progress bar based on the progress of webpack reading the file

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/webpack-builder/package.json [hunk 1]
```diff
@@ -93,7 +93,11 @@
     "style-loader": "3.3.1",
     "terser-webpack-plugin": "5.3.3",
     "ts-loader": "9.3.1",
-    "webpack": "5.73.0"
+    "webpack": "5.73.0",
+    "chalk": "4.1.0",
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/package.json:97]: chalk 已经预打包了一份，可以直接引用 @modern-js/utils/chalk 哈，就不用额外安装了

File: packages/builder/webpack-builder/src/plugins/progress.ts [hunk 2]
```diff
@@ -20,3 +33,148 @@ export const PluginProgress = (): BuilderPlugin => ({
...
+
+export interface IProgressOptions
+  extends Omit<Props, 'message' | 'total' | 'current' | 'done'> {
+  quiet: boolean;
+  quietOnDev: boolean;
+  id: string;
+  clearOnDone: boolean;
+}
+
+export const UseWebpackProgressPlugin = (
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/plugins/progress.ts:56]: 建议把 webpack 插件放到 webpackPlugins 目录下面

File: packages/builder/webpack-builder/src/plugins/progress.ts [hunk 3]
```diff
@@ -20,3 +33,148 @@ export const PluginProgress = (): BuilderPlugin => ({
...
+  name: 'webpack-builder-plugin-progress',
+  setup(api) {
+    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
+      const config = api.getBuilderConfig();
+      if (!config.dev?.progressBar) {
+        return;
+      }
+      chain
+        .plugin(CHAIN_ID.PLUGIN.PROGRESS)
+        .use(UseWebpackProgressPlugin, [cwd, options]);
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/plugins/progress.ts:177]: UseWebpackProgressPlugin 是一个函数，不是一个 webpack 插件，应该没法直接放在 chain.use 里使用

Comment (chenjiahan) [reply packages/builder/webpack-builder/src/plugins/progress.ts:177]: 可以在 modern.js/tests/integration/webpack-builder 跑 dev 和 build 命令看一下集成后的效果哈

File: packages/builder/webpack-builder/src/plugins/progress.ts (hunk 2 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/src/plugins/progress.ts:56]: ok

File: packages/builder/webpack-builder/package.json (hunk 1 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/package.json:97]: ok

File: packages/builder/webpack-builder/src/plugins/progress.ts (hunk 3 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/src/plugins/progress.ts:177]: ok

**Comment by chenjiahan** [comment]
main 分支调整了一下 builder npm 包的结构，需要重新 merge 一下最新代码~

File: packages/builder/webpack-builder/src/plugins/progress.ts (hunk 3 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/src/plugins/progress.ts:177]: image

File: packages/builder/webpack-builder/src/plugins/progress.ts (hunk 2 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/src/plugins/progress.ts:56]: image

File: packages/builder/webpack-builder/package.json (hunk 1 above)
Comment (zhusjfaker) [reply packages/builder/webpack-builder/package.json:97]: image

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+      }
+      if (quietOnDev && process.env.NODE_ENV === 'development') {
+        return;
+      }
+      /**
+       * Cache tool files
+       */
+      const totalFilePath = resolve(
+        os.tmpdir(),
+        `./pia-plugin-progress_${getHash(`${id}_${cwd}`, null as any)}`,
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:57]: 这里写到磁盘上的目的是啥

如果要写到磁盘的话，建议放在 node_modules/.cache/builder 目录下，然后路径名字里不要包含 pia

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+      compiler.hooks.compilation.tap(pluginName, compilation => {
+        const hooks =
+          compiler.webpack.NormalModule.getCompilationHooks(compilation);
+
+        hooks.readResource
+          .for(undefined)
+          .tapPromise('onLoad', async (loaderContext: any) => {
+            progress.loaded += 1;
+            // loaderContext.addDependency(loaderContext.resourcePath);
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:83]: 注释代码可以删掉哈

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+export const UseWebpackProgressPlugin = (
+  cwd: string,
+  options: Partial<IProgressOptions>,
+): webpack.WebpackPluginInstance => {
+  let total = 0;
+  const {
+    id = 'Modern',
+    quiet = false,
+    quietOnDev = true,
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:37]: quietOnDev 为啥默认是 true 嘞，dev 时是需要进度条的

File: packages/builder/builder-webpack-provider/src/plugins/progress.ts [hunk 7]
```diff
@@ -20,3 +34,19 @@ export const PluginProgress = (): BuilderPlugin => ({
...
   },
 });
+
+export const PluginProgress = (
+  cwd: string = process.cwd(),
+  options: Partial<IProgressOptions> = {},
+): BuilderPlugin => ({
+  name: 'webpack-builder-plugin-progress',
+  setup(api) {
+    api.modifyWebpackConfig(webpack_config => {
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/plugins/progress.ts:44]: 变量名统一用驼峰哈

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts (hunk 5 above)
Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:83]: ok

Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:83]: image

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts (hunk 6 above)
Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:37]: image

File: packages/builder/builder-webpack-provider/src/plugins/progress.ts (hunk 7 above)
Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/plugins/progress.ts:44]: image

File: packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts (hunk 4 above)
Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:57]: image

Comment (zhusjfaker) [reply packages/builder/builder-webpack-provider/src/webpackPlugins/ProgressPlugin.ts:57]: node_modules 经常删除和修改 不同的 包处理器 路径获取也不太一致，对进度条的增量 和 处理 可能产生问题 这里文件位置 就暂时不改了 名称我已经改过了

**Review by chenjiahan — DISMISSED** [review]

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**